src/entry.c

```diff
@@ -0,0 +1,385 @@
+#include <stdbool.h>
+#include "server.h"
+#include "serverassert.h"
+#include "entry.h"
+
+#include <stdbool.h>
+
+/*-----------------------------------------------------------------------------
+ * Entry API
+ *----------------------------------------------------------------------------*/
+
+/* The entry pointer is the field sds. We encode the entry layout type
+ * in the field SDS header. Field type SDS_TYPE_5 doesn't have any spare bits to
+ * encode this so we use it only for the first layout type.
+ *
+ * Entry with embedded value, used for small sizes. The value is stored as
+ * SDS_TYPE_8. The field can use any SDS type.
+ *
+ * Entry can also have expiration timestamp, which is the UNIX timestamp for it to be expired.
+ * For aligned fast access, we keep the expiry timestamp prior to the start of the sds header.
+ *
+ *     +--------------+--------------+---------------+
+ *     | Expiration   | field        | value         |
+ *     | 1234567890LL | hdr "foo" \0 | hdr8 "bar" \0 |
+ *     +--------------+--------------+---------------+
+ *
+ * Entry with value pointer, used for larger fields and values. The field is SDS
+ * type 8 or higher.
+ *
+ *     +--------------+-------+--------------+
+ *     | Expiration   | value | field        |
+ *     | 1234567890LL | ptr   | hdr "foo" \0 |
+ *     +--------------+---^---+--------------+
+ *                        |
+ *                        |
+ *                        value pointer = value sds
+ */
+
+/* SDS aux flag. If set, it indicates that the entry has TTL metadata set. */
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY 0
+
+/* SDS aux flag. If set, it indicates that the entry has an embedded value
+ * pointer located in memory before the embedded field. If unset, the entry
+ * instead has an embedded value located after the embedded field. */
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR 2
```

Comment:
consider turning them into enum and using `static_assert` to express the contract that they can go beyond 4, inclusive. 

```suggestion
enum {
    /* SDS aux flag. If set, it indicates that the entry has TTL metadata set. */
    FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY = 0,

    /* SDS aux flag. If set, it indicates that the entry has an embedded value
     * pointer located in memory before the embedded field. If unset, the entry
     * instead has an embedded value located after the embedded field. */
    FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR = 2,
    FIELD_SDS_AUX_BIT_MAX
};
static_assert(FIELD_SDS_AUX_BIT_MAX < 8-SDS_TYPE_BITS);
```